### ISSUE TITLE
app.initialize() promise never resolves or rejects, when not hosted in Teams (or M365 app)

### ISSUE THREAD

**Issue by wictorwilen** [OfficeDev/microsoft-teams-library-js#719]
I expect the `app.initialize` method in the promise based SDK v2 to return either a resolved promise - when hosted in Microsoft Teams or an M365 app, or a rejected value/error - when hosted outside any Teams/M365 context.

The current implementation never resolves (or rejects) the Promise, which is not a good practice.

Example to reproduce: Create an HTML file with the following content and observe that nothing happens:

``` html
<html>

<head></head>

<body>
    <script src="https://res.cdn.office.net/teams-js/2.0.0-beta.1/js/MicrosoftTeams.min.js"
        integrity="sha384-tQzNKNgoXVyO8/7UVmwwNjKy8Cai2COqlhP7/hL4Jz5flCpugVVTzNWo3SnGEpH4"
        crossorigin="anonymous"></script>

    <script>
        microsoftTeams.
            app.
            initialize().
            then(() => { alert("Initialized") }).
            catch(err => { alert("Not initialized") });

    </script>
</body>

</html>
```

**Comment by v-chetsh** [comment]
@wictorwilen - We will check on it from our end also and let you know.

**Comment by MarcinJuraszek** [comment]
Thank you for reporting the issue, @wictorwilen. This is somewhat related to another issue that's been files a while ago (and you've been part of the discussion on), which is how to detect if you're running in Teams. See https://github.com/OfficeDev/microsoft-teams-library-js/issues/469

I think the simplest thing to do here, just to unblock you and other people running into this right now, will be adding a timeout to the `initialize` implementation. It would result in the promise being rejected if we don't hear back from Teams/Office/Outlook in some pre-defined time (couple seconds?). 

Long term we'll need to figure out a better way to fail right away if not running in Teams (or Office or Outlook), without the delay.

**Comment by wictorwilen** [comment]
Thanks @MarcinJuraszek. I think a timeout in the SDK is the current best way to solve this. I experimented with a timeout wrapper in my Teams JS SDK wrapper, but a "native" one makes more sense. Also, they always unreliable way of checking the iframe host is even more unreliable now with office.com not setting an iframe name etc etc

**Comment by MarcinJuraszek** [comment]
Yes, using something the host (Teams, Office, Outlook, etc.) set on the iframe would be possible, but it's very fragile. Especially when you consider multiple platforms (web, desktop, ios, android) too. I'll work on adding a timeout for now. We'll need to think of a more scalable solution later.

**Comment by wictorwilen** [comment]
Looking forward to it!

**Comment by marekcis** [comment]
Our clients has also started to having issues with Teams SDK loading.
We've already wrapped Teams initialise method in Promise with timeout, something like this:
```
initialize(): Promise<void> {
      return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
          reject(new SDKTimeoutError());
          return;
        }, 10000);

        microsoftTeams.initialize(() => {
          clearTimeout(timeout);
          resolve();
        });
      });
    }
```

Normally it was 5 seconds, I've changed it to 10 sec. 
I would like to know what is the root cause of this issue.  
It's not completely clear for me, why this issue is connected with [embed detecting](https://github.com/OfficeDev/microsoft-teams-library-js/issues/469) in Teams.

What would be next steps we need to make in order to have `app.initialize()` running as usual?
Thanks for your investigation and help.

**Comment by MarcinJuraszek** [comment]
> It's not completely clear for me, why this issue is connected with embed detecting in Teams.

In the ideal scenario, the `initialize` function would first check if it's running in Teams/Office/Outlook and reject immediately if that's not the case. 

If we don't know if we're running in Teams/Office/Outlook, there isn't much more we can do than wait for the `initialize` response to come back or time out at some point and fail the call.

**Comment by marekcis** [comment]
@MarcinJuraszek thanks for clarifying. And what timeout do you suggest for 3rd party developers to implement?

**Comment by Prasad-MSFT** [comment]
@MarcinJuraszek - could you please help to confirm the recommendation for 3P developers?

**Comment by MarcinJuraszek** [comment]
I'm sorry, I'm not familiar with the scenario you described around "Teams meeting on custom domains". The case I had in mind is a website that's deployed in a way that allows for it to be accessed and used directly OR as a Teams Tab. In this case the site would use TeamsJs, but when it's accessed directly it would never be able to finish initialization (because it's not running as a Teams Tab). In this case the Promise returned from `initialization` would never resolve or reject. I recently made a change to 2.0-preview branch to time out that `initialization` API after 5 seconds. To be honest, I don't really have a good reason why it's 5, and not 3 or 10 seconds. I hope at some point next year we'll invest in having a different way of finding out if we're running in a Teams Tab context and if not we would reject the promise right away.

**Comment by marekcis** [comment]
thanks for clarifying :) Will look forward to upcoming findings / changes in initialisation next year.

**Comment by v-chetsh** [comment]
We are closing this issue for now, Please feel free to reopen it, if required.
